htmresearch/frameworks/poirazi_neuron_model/neuron_model.py

```diff
@@ -0,0 +1,231 @@
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import random
+import numpy
+import copy
+from sklearn.cluster import AgglomerativeClustering
+from collections import Counter
+from nupic.bindings.math import *
+
+numpy.set_printoptions(threshold=numpy.inf)
+
+def power_nonlinearity(power):
+    def l(activations):
+        original_activations = copy.deepcopy(activations)
+        for i in range(power - 1):
+            activations.elementNZMultiply(original_activations)
+        return activations
+    return l
+
+def threshold_nonlinearity(threshold):
+    def l(activations):
+        activations.threshold(threshold)
+        return activations
+    return l
+
+def sigmoid(center, scale):
+	return lambda x: 1./(1. + numpy.exp(scale*(center - x)))
+
+def sigmoid_nonlinearity(center, scale):
+	def l(activations):
+		dense = activations.toDense()
+		f = sigmoid(center, scale)
+		return SM32(f(dense))
+	return l
+
+class Matrix_Neuron(object):
+    def __init__(self,
+                 size = 10000,
+                 num_dendrites = 1000,
+                 dendrite_length = 10,
+                 dim = 400,
+                 nonlinearity = threshold_nonlinearity(6),
+                 num_replacement = 25,
+                 initial_permanence = 0.5,
+                 permanence_threshold = 0.15,
+                 perm_dec = 0.0125,
+                 perm_inc = 0.02):
+        self.size = size
+        self.num_dendrites = num_dendrites
+        self.dendrite_length = dendrite_length
+        self.dim = dim
+        self.nonlinearity = nonlinearity
+        self.num_replacement = num_replacement
+        self.initial_permanence = initial_permanence
+        self.permanence_threshold = permanence_threshold
+        self.perm_dec = perm_dec
+        self.perm_inc = perm_inc
+        self.initialize_dendrites()
+        self.initialize_permanences()
+
+    def initialize_dendrites(self):
+        """
+        Initialize all the dendrites of the neuron to a set of random connections
+        """
+        # Wipe any preexisting connections by creating a new connection matrix
+        self.dendrites = SM32()
+        self.dendrites.reshape(self.dim, self.num_dendrites)
+
+        for row in range(self.num_dendrites):
+            synapses = numpy.random.choice(self.dim, self.dendrite_length, replace = False)
+            for synapse in synapses:
+                self.dendrites[synapse, row] = 1
+
+
+    def initialize_permanences(self):
+        self.permanences = copy.deepcopy(self.dendrites)
+        self.permanences = self.permanences*0.3
+
+    def calculate_activation(self, datapoint):
+        """
+        Only for a single datapoint
+        """
+
+        activations = datapoint * self.dendrites
+        activations = self.nonlinearity(activations)
+        return activations.sum()
+
+    def choose_substitute_pool(self):
```

Comment:
Oops, it's not.  That's a remnant of the old P&M learning rule implementation, which I thought I had completely scrubbed from this code.